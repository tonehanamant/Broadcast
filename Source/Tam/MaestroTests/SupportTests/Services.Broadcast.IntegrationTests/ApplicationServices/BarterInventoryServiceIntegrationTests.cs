using ApprovalTests;
using ApprovalTests.Reporters;
using Common.Services;
using IntegrationTests.Common;
using Microsoft.Practices.Unity;
using Newtonsoft.Json;
using NUnit.Framework;
using Services.Broadcast.ApplicationServices;
using Services.Broadcast.ApplicationServices.Security;
using Services.Broadcast.Entities;
using Services.Broadcast.Entities.BarterInventory;
using Services.Broadcast.Entities.StationInventory;
using Services.Broadcast.Exceptions;
using Services.Broadcast.IntegrationTests.Stubbs;
using Services.Broadcast.Repositories;
using System;
using System.Collections.Generic;
using System.IO;
using Tam.Maestro.Common.DataLayer;
using Tam.Maestro.Data.Entities;
using Tam.Maestro.Services.Cable.SystemComponentParameters;

namespace Services.Broadcast.IntegrationTests.ApplicationServices
{
    [TestFixture]
    public class BarterInventoryServiceIntegrationTests
    {
        private IBarterInventoryService _BarterService = IntegrationTestApplicationServiceFactory.GetApplicationService<IBarterInventoryService>();
        private IInventoryFileRepository _InventoryFileRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IInventoryFileRepository>();
        private IInventoryRepository _IInventoryRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IInventoryRepository>();
        private IBarterRepository _BarterRepository = IntegrationTestApplicationServiceFactory.BroadcastDataRepositoryFactory.GetDataRepository<IBarterRepository>();     

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SaveBarterInventoryFile()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_ValidFormat.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                var jsonResolver = new IgnorableSerializerContractResolver();
                jsonResolver.Ignore(typeof(InventoryFileBase), "Id");
                var jsonSettings = new JsonSerializerSettings()
                {
                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                    ContractResolver = jsonResolver
                };

                var file = _InventoryFileRepository.GetInventoryFileById(result.FileId);
                var fileJson = IntegrationTestHelper.ConvertToJson(file, jsonSettings);

                Approvals.Verify(fileJson);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SavesManifests_SingleBook()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_ValidFormat_SingleBook.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SavesManifests_TwoBooks()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_ValidFormat_TwoBooks.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SavesManifests_WhenStationIsUnknown()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_ValidFormat_UnknownStation.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_ValidationErrors()
        {
            const string fileName = @"BarterDataFiles\Barter DataLines file with missed values.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };
                var problems = new List<InventoryFileProblem>();

                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_InvalidFileFormat()
        {
            const string fileName = @"1Chicago WLS Syn 4Q16 UNKNOWN.xml";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryFileSaveResult(result);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterFileImporter_ExtractData_PRI5390()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_BadFormats_PRI5390.xlsx";
            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SaveBarterInventoryFile_PRI5667()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_BadFormats_PRI5667.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SaveBarterInventoryFile_InvalidDaypart()
        {
            const string fileName = @"BarterDataFiles\Barter DataLines file with invalid daypart.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SaveBarterInventoryFile_SingleDataColumn()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_SingleDataColumn.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SaveBarterInventoryFile_DateRangeIntersecting()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_DateRangeIntersecting.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void OAndO_NotValidFile1()
        {
            const string fileName = @"BarterDataFiles\OAndO_InvalidFile1.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void OAndO_NotValidFile2()
        {
            const string fileName = @"BarterDataFiles\OAndO_InvalidFile2.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void OAndO_NotValidFile3()
        {
            const string fileName = @"BarterDataFiles\OAndO_InvalidFile3.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void OAndO_NotValidFile4()
        {
            const string fileName = @"BarterDataFiles\OAndO_InvalidFile4.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var problems = new List<InventoryFileProblem>();
                try
                {
                    var now = new DateTime(2019, 02, 02);
                    var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);
                }
                catch (FileUploadException<InventoryFileProblem> e)
                {
                    problems = e.Problems;
                }

                _VerifyInventoryFileProblems(problems);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void SavesOAndOBarterInventoryFile()
        {
            const string fileName = @"BarterDataFiles\OAndO_ValidFile1.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                var jsonResolver = new IgnorableSerializerContractResolver();
                jsonResolver.Ignore(typeof(InventoryFileBase), "Id");
                jsonResolver.Ignore(typeof(BarterInventoryFile), "CreatedDate");
                jsonResolver.Ignore(typeof(InventorySource), "Id");
                var jsonSettings = new JsonSerializerSettings()
                {
                    ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                    ContractResolver = jsonResolver
                };

                var file = _BarterRepository.GetBarterInventoryFileById(result.FileId);
                var fileJson = IntegrationTestHelper.ConvertToJson(file, jsonSettings);

                Approvals.Verify(fileJson);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void BarterInventoryService_SendFileToDataLake()
        {
            const string fileName = @"BarterDataFiles\BarterFileImporter_ValidFormat.xlsx";

            using (new TransactionScopeWrapper())
            {
                IntegrationTestApplicationServiceFactory.Instance.RegisterType<IImpersonateUser, ImpersonateUserStubb>();
                IntegrationTestApplicationServiceFactory.Instance.RegisterType<IFileService, FileServiceDataLakeStubb>();
                var fileService = IntegrationTestApplicationServiceFactory.Instance.Resolve<IFileService>();
                var dataLakeFolder = BroadcastServiceSystemParameter.DataLake_SharedFolder;
                var filePath = Path.Combine(dataLakeFolder, fileName);
                var barterService = IntegrationTestApplicationServiceFactory.GetApplicationService<IBarterInventoryService>();

                if (fileService.Exists(filePath))
                {
                    fileService.Delete(filePath);
                }

                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = barterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                Assert.True(fileService.Exists(filePath));
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void SaveBarterInventoryFile_DuplicateDataError_ForecastDb_TwoBooks()
        {
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IImpersonateUser, ImpersonateUserStubb>();
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IFileService, FileServiceDataLakeStubb>();
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IEmailerService, EmailerServiceStubb>();

            const string fileName = @"BarterDataFiles\DuplicateData_ForecastDB_TwoBooks.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        [Test]
        [UseReporter(typeof(DiffReporter))]
        public void SaveBarterInventoryFile_DuplicateDataError_ForecastDb_SingleBook()
        {
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IImpersonateUser, ImpersonateUserStubb>();
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IFileService, FileServiceDataLakeStubb>();
            IntegrationTestApplicationServiceFactory.Instance.RegisterType<IEmailerService, EmailerServiceStubb>();

            const string fileName = @"BarterDataFiles\DuplicateData_ForecastDB_SingleBook.xlsx";

            using (new TransactionScopeWrapper())
            {
                var request = new InventoryFileSaveRequest
                {
                    StreamData = new FileStream($@".\Files\{fileName}", FileMode.Open, FileAccess.Read),
                    FileName = fileName
                };

                var now = new DateTime(2019, 02, 02);
                var result = _BarterService.SaveBarterInventoryFile(request, "IntegrationTestUser", now);

                _VerifyInventoryGroups(result.FileId);
            }
        }

        private static void _VerifyInventoryFileProblems(List<InventoryFileProblem> problems)
        {
            var jsonResolver = new IgnorableSerializerContractResolver();
            var jsonSettings = new JsonSerializerSettings()
            {
                ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                ContractResolver = jsonResolver
            };
            var problemsJson = IntegrationTestHelper.ConvertToJson(problems, jsonSettings);

            Approvals.Verify(problemsJson);
        }

        private static void _VerifyInventoryFileSaveResult(InventoryFileSaveResult result)
        {
            var jsonResolver = new IgnorableSerializerContractResolver();
            var jsonSettings = new JsonSerializerSettings()
            {
                ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                ContractResolver = jsonResolver
            };

            Approvals.Verify(IntegrationTestHelper.ConvertToJson(result, jsonSettings));
        }

        private void _VerifyInventoryGroups(int fileId)
        {
            var jsonResolver = new IgnorableSerializerContractResolver();
            jsonResolver.Ignore(typeof(StationInventoryGroup), "Id");
            jsonResolver.Ignore(typeof(StationInventoryManifestBase), "Id");
            jsonResolver.Ignore(typeof(StationInventoryManifestBase), "FileId");
            jsonResolver.Ignore(typeof(StationInventoryManifestAudience), "Id");
            jsonResolver.Ignore(typeof(StationInventoryManifestWeek), "Id");
            jsonResolver.Ignore(typeof(StationInventoryManifestDaypart), "Id");
            jsonResolver.Ignore(typeof(StationInventoryManifestRate), "Id");
            jsonResolver.Ignore(typeof(MediaWeek), "_Id");
            jsonResolver.Ignore(typeof(DisplayBroadcastStation), "Id");
            var jsonSettings = new JsonSerializerSettings()
            {
                ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
                ContractResolver = jsonResolver
            };

            var groups = _IInventoryRepository.GetStationInventoryGroupsByFileId(fileId);
            var groupsJson = IntegrationTestHelper.ConvertToJson(groups, jsonSettings);

            Approvals.Verify(groupsJson);
        }
    }
}
