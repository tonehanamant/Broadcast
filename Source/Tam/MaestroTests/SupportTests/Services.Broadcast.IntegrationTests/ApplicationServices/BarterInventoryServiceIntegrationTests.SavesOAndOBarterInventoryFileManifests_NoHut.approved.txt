[
  {
    "Comment": null,
    "DaypartCode": "EMN",
    "EffectiveDate": "2019-04-01T00:00:00",
    "EndDate": "2019-04-02T00:00:00",
    "InventorySourceId": 10,
    "ManifestAudiences": [
      {
        "Audience": {
          "AudienceString": "House Holds",
          "Id": 31
        },
        "CPM": 6.2500,
        "Impressions": 160000.0,
        "IsReference": false,
        "Rating": null
      }
    ],
    "ManifestAudiencesReferences": [],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 35999,
          "_Friday": false,
          "_Id": 51352,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": false,
          "_Name": "Custom",
          "_Saturday": true,
          "_StartTime": 21600,
          "_Sunday": true,
          "_Thursday": false,
          "_Tuesday": false,
          "_Wednesday": false
        },
        "Genres": [],
        "ProgramName": "Eyewitness News/GMA Weekend"
      }
    ],
    "ManifestRates": [],
    "ManifestWeeks": [
      {
        "MediaWeek": {
          "_EndDate": "2019-02-10T00:00:00",
          "_MediaMonthId": 449,
          "_StartDate": "2019-02-04T00:00:00",
          "_WeekNumber": 2,
          "IsDirty": true
        },
        "Spots": 5
      },
      {
        "MediaWeek": {
          "_EndDate": "2019-02-24T00:00:00",
          "_MediaMonthId": 449,
          "_StartDate": "2019-02-18T00:00:00",
          "_WeekNumber": 4,
          "IsDirty": true
        },
        "Spots": 5
      },
      {
        "MediaWeek": {
          "_EndDate": "2019-03-03T00:00:00",
          "_MediaMonthId": 450,
          "_StartDate": "2019-02-25T00:00:00",
          "_WeekNumber": 1,
          "IsDirty": true
        },
        "Spots": 5
      }
    ],
    "ProjectedStationImpressions": [],
    "SpotLengthId": 3,
    "SpotsPerDay": null,
    "SpotsPerWeek": null,
    "Station": {
      "Affiliation": "ABC",
      "CallLetters": "WABC-TV 7.1",
      "Code": 5015,
      "LegacyCallLetters": "WABC",
      "ManifestMaxEndDate": null,
      "MarketCode": 101,
      "ModifiedDate": null,
      "OriginMarket": null,
      "RateDataThrough": null
    }
  },
  {
    "Comment": null,
    "DaypartCode": "EMN",
    "EffectiveDate": "2019-04-01T00:00:00",
    "EndDate": "2019-04-02T00:00:00",
    "InventorySourceId": 10,
    "ManifestAudiences": [
      {
        "Audience": {
          "AudienceString": "House Holds",
          "Id": 31
        },
        "CPM": 5.6250,
        "Impressions": 160000.0,
        "IsReference": false,
        "Rating": null
      }
    ],
    "ManifestAudiencesReferences": [],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 35999,
          "_Friday": false,
          "_Id": 51352,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": false,
          "_Name": "Custom",
          "_Saturday": true,
          "_StartTime": 21600,
          "_Sunday": true,
          "_Thursday": false,
          "_Tuesday": false,
          "_Wednesday": false
        },
        "Genres": [],
        "ProgramName": "Eyewitness News/GMA Weekend"
      }
    ],
    "ManifestRates": [],
    "ManifestWeeks": [
      {
        "MediaWeek": {
          "_EndDate": "2019-02-10T00:00:00",
          "_MediaMonthId": 449,
          "_StartDate": "2019-02-04T00:00:00",
          "_WeekNumber": 2,
          "IsDirty": true
        },
        "Spots": 2
      },
      {
        "MediaWeek": {
          "_EndDate": "2019-03-03T00:00:00",
          "_MediaMonthId": 450,
          "_StartDate": "2019-02-25T00:00:00",
          "_WeekNumber": 1,
          "IsDirty": true
        },
        "Spots": 2
      }
    ],
    "ProjectedStationImpressions": [],
    "SpotLengthId": 3,
    "SpotsPerDay": null,
    "SpotsPerWeek": null,
    "Station": {
      "Affiliation": "ABC",
      "CallLetters": "WABC-TV 7.1",
      "Code": 5015,
      "LegacyCallLetters": "WABC",
      "ManifestMaxEndDate": null,
      "MarketCode": 101,
      "ModifiedDate": null,
      "OriginMarket": null,
      "RateDataThrough": null
    }
  },
  {
    "Comment": null,
    "DaypartCode": "EMN",
    "EffectiveDate": "2019-04-01T00:00:00",
    "EndDate": "2019-04-02T00:00:00",
    "InventorySourceId": 10,
    "ManifestAudiences": [
      {
        "Audience": {
          "AudienceString": "House Holds",
          "Id": 31
        },
        "CPM": 6.2500,
        "Impressions": 160000.0,
        "IsReference": false,
        "Rating": null
      }
    ],
    "ManifestAudiencesReferences": [],
    "ManifestDayparts": [
      {
        "Daypart": {
          "_Code": "CUS",
          "_EndTime": 35999,
          "_Friday": false,
          "_Id": 51352,
          "_Is24Hour": false,
          "_LastEndTime": 0,
          "_LastStartTime": 0,
          "_Monday": false,
          "_Name": "Custom",
          "_Saturday": true,
          "_StartTime": 21600,
          "_Sunday": true,
          "_Thursday": false,
          "_Tuesday": false,
          "_Wednesday": false
        },
        "Genres": [],
        "ProgramName": "Eyewitness News/GMA Weekend"
      }
    ],
    "ManifestRates": [],
    "ManifestWeeks": [
      {
        "MediaWeek": {
          "_EndDate": "2019-02-17T00:00:00",
          "_MediaMonthId": 449,
          "_StartDate": "2019-02-11T00:00:00",
          "_WeekNumber": 3,
          "IsDirty": true
        },
        "Spots": 1
      },
      {
        "MediaWeek": {
          "_EndDate": "2019-03-03T00:00:00",
          "_MediaMonthId": 450,
          "_StartDate": "2019-02-25T00:00:00",
          "_WeekNumber": 1,
          "IsDirty": true
        },
        "Spots": 2
      }
    ],
    "ProjectedStationImpressions": [],
    "SpotLengthId": 3,
    "SpotsPerDay": null,
    "SpotsPerWeek": null,
    "Station": {
      "Affiliation": "ABC",
      "CallLetters": "WABC-TV 7.1",
      "Code": 5015,
      "LegacyCallLetters": "WABC",
      "ManifestMaxEndDate": null,
      "MarketCode": 101,
      "ModifiedDate": null,
      "OriginMarket": null,
      "RateDataThrough": null
    }
  }
]